ethdb/privateapi/ethbackend.go

```diff
@@ -679,6 +680,91 @@ func (s *EthBackendServer) EngineForkChoiceUpdated(ctx context.Context, req *rem
 	}, nil
 }
 
+func (s *EthBackendServer) EngineGetPayloadBodiesByHashV1(ctx context.Context, request *remote.EngineGetPayloadBodiesByHashV1Request) (*remote.EngineGetPayloadBodiesV1Response, error) {
+	tx, err := s.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bodies := make([]*types2.ExecutionPayloadBodyV1, len(request.Hashes))
+
+	for hashIdx, hash := range request.Hashes {
+		h := gointerfaces.ConvertH256ToHash(hash)
+		block, err := rawdb.ReadBlockByHash(tx, h)
+		if err != nil {
+			return nil, err
+		}
+
+		body, err := extractPayloadBodyFromBlock(block)
+		if err != nil {
+			return nil, err
+		}
+		bodies[hashIdx] = body
+	}
+
+	return &remote.EngineGetPayloadBodiesV1Response{Bodies: bodies}, nil
+}
+
+func (s *EthBackendServer) EngineGetPayloadBodiesByRangeV1(ctx context.Context, request *remote.EngineGetPayloadBodiesByRangeV1Request) (*remote.EngineGetPayloadBodiesV1Response, error) {
+	tx, err := s.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bodies := make([]*types2.ExecutionPayloadBodyV1, request.Count)
+
+	var i uint64
+	for i = 0; i < request.Count; i++ {
+		block, err := rawdb.ReadBlockByNumber(tx, request.Start+i)
```

Comment:
Trailing null blocks (blocks after the latest available) should be truncated (see Item 3 of [the spec](https://github.com/ethereum/execution-apis/blob/main/src/engine/shanghai.md#specification-4)). In our database we don't allow any gaps in canonical blocks, so it's OK to truncate starting from the first missing block.